components/script/dom/offscreencanvasrenderingcontext2d.rs

```diff
@@ -0,0 +1,306 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding;
+use dom::bindings::codegen::Bindings::OffscreenCanvasRenderingContext2DBinding::OffscreenCanvasRenderingContext2DMethods;
+use dom::bindings::reflector::{Reflector, reflect_dom_object};
+use dom::bindings::root::{Dom,DomRoot};
+use dom::globalscope::GlobalScope;
+use dom::offscreencanvas::{OffscreenCanvas};
+use dom_struct::dom_struct;
+use euclid::Size2D;
+
+#[dom_struct]
+pub struct OffscreenCanvasRenderingContext2D{
+    reflector_: Reflector,
+    canvas: Option<Dom<OffscreenCanvas>>,
+}
+
+impl OffscreenCanvasRenderingContext2D {
+    pub fn new_inherited(
+        global: &GlobalScope,
+        canvas: Option<&OffscreenCanvas>,
+        size: Size2D<u64>,
+    ) -> OffscreenCanvasRenderingContext2D {
+        OffscreenCanvasRenderingContext2D {
+            reflector_: Reflector::new(),
+            canvas: canvas.map(Dom::from_ref),
+        }
+    }
+
+    pub fn new(
+        global: &GlobalScope,
+        canvas: &OffscreenCanvas,
+        size: Size2D<u64>,
+    ) -> DomRoot<OffscreenCanvasRenderingContext2D> {
+        //let window = window_from_node(canvas);
+        let boxed = Box::new(OffscreenCanvasRenderingContext2D::new_inherited(
+            global,
+            Some(canvas),
+            size,
+        ));
+        reflect_dom_object(boxed, global, OffscreenCanvasRenderingContext2DBinding::Wrap)
+    }
+}
+
+impl OffscreenCanvasRenderingContext2DMethods for OffscreenCanvasRenderingContext2D {
+
+    fn Canvas(&self) -> DomRoot<OffscreenCanvas> {
+        DomRoot::from_ref(self.canvas.as_ref().expect("No canvas."))
+    }
+
+    // ***************** For Mixin CanvasState ***********************************
+	// https://html.spec.whatwg.org/multipage/#dom-context-2d-save
+	fn Save(&self) {
+		self.saved_states
```

Comment:
This doesn't compile, right? Rather than duplicating all of the canvas implementation code, I wanted to move the relevant implementation of CanvasRenderingContext2D methods to a structure that can be shared by both CanvasRenderingContext2D and OffscreenCanvasRenderingContext2D. For example:
```rust
struct CanvasState {
    saved_states: ...,
    ...
}
impl CanvasState {
    fn save(&self) {
        ....
    }
}

impl OffscreenCanvasRenderingContext2DMethods for OffscreenCanvasRenderingContext2D {
    fn Save(&self) {
        self.canvas_state.save()
    }
}

impl CanvasRenderingContext2DMethods for CanvasRenderingContext2D {
    fn Save(&self) {
        self.canvas_state.save()
    }
}
```